src/librustc/ty/structural_impls.rs

```diff
@@ -835,6 +876,36 @@ impl<'tcx> TypeFoldable<'tcx> for ty::instance::Instance<'tcx> {
             },
         }
     }
+
+    fn super_hash_with<H: TypeHasher<'tcx>>(&self, hasher: &mut H) -> u64 {
+        use ty::InstanceDef::*;
+        self.substs.hash_with(hasher);
+        match self.def {
```

Comment:
these hasher impls don't quite look right. We need to hash `mem::discriminant(self)` as well. Here is a good example:

https://github.com/rust-lang/rust/blob/ed0350e94569d74bdfe0e612fedfd5ab173f5242/src/librustc/ich/impls_mir.rs#L38-L54

Reply:
Ah, got it - was worrying about the fact that I was hashing a 'flattened' structure rather than the tree structure but I wasn't entirely sure how to resolve it. Presumably this is effectively adding in the enum 'position' to the hash